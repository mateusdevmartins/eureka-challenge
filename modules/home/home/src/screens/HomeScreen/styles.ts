import styled from 'styled-components';

export const LogoImage = styled.img`
  align-self: left;
  justify-content: 'flex-start';
  margin: 0px 8px 0px 8px;
  width: 144px;
  heigth: 36px;
  padding: 24px 0px 24px 72px;
`;

export const MainView = styled.div`
  display: flex;
  flex: 1;
  flex-direction: column;
  background-repeat: initial;
`;

export const ContentView = styled.div`
  display: flex;
  justify-content: flex-start;
  flex-direction: row;
  padding: 67px 22px 67px 22px;
  @media (max-width: 1100px) {
    flex-direction: column;
    align-items: center;
    padding: 67px 22px 0px 22px;
  }
`;

export const LogoView = styled.div`
  width: 100%;
  background-color: white;
`;
