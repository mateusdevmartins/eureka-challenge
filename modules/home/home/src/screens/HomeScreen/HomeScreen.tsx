/* eslint-disable react-hooks/exhaustive-deps */
import { useCallback, useEffect, useState } from 'react';
import {
  authorize,
  getProgramsById,
  getProgramsLevelsById,
  getUserActivitiesById,
  getUserProfileById,
  getUsers,
} from '@fieloapp/services';
import { Activity, ProgramLevels, Programs, User } from '@fieloapp/homestore';
import { setItem } from '@fieloapp/storage';
import { images } from '@fieloapp/uikit';
import { UserActivities, UserProfile, UserRanking } from '../../components';
import { MainView, LogoImage, ContentView, LogoView } from './styles';

const { logo } = images;

export function HomeScreen() {
  const [usersState, setUsersState] = useState<Array<User>>([]);
  const [selectedUser, setSelectedUser] = useState<User>();
  const [selectedProgram, setSelectedProgram] = useState<Programs>();
  const [programLevels, setProgramLevels] = useState<Array<ProgramLevels>>();
  const [backgroundImage, setBackgroundImage] = useState<string>();
  const [userActivities, setUserActivities] = useState<Array<Activity>>();

  const getAllUsers = useCallback(async () => {
    const response = await getUsers();
    if (response?.data) {
      const user = [...response.data];
      user.sort((userA, userB) => {
        return userB.balance.points - userA.balance.points;
      });

      setUsersState(user);
    }
  }, []);

  const getAuthorization = useCallback(async () => {
    const response = await authorize();
    if (response?.data?.auth === true) {
      await setItem('auth', response.data.token);
      getAllUsers();
    }
  }, []);

  const getWallpaper = useCallback(async () => {
    const ajaxRequest = new XMLHttpRequest();
    ajaxRequest.open(
      'POST',
      'http://www.bing.com/HPImageArchive.aspx?format=xml&idx=0&n=1&mkt=en-US',
      true
    );

    ajaxRequest.send('');
    ajaxRequest.onreadystatechange = function () {
      if (ajaxRequest.readyState === 4) {
        const res = this.response.split('<url>');
        const res1 = res[1].split('</url>');
        setBackgroundImage('https://bing.com' + res1[0]);
      }
    };
  }, []);

  useEffect(() => {
    getAuthorization();
    getWallpaper();
  }, []);

  const selectUser = (user: User) => {
    getUserProfile(user.id);
  };

  const getProgramsLevels = useCallback(async (programId: string) => {
    const response = await getProgramsLevelsById(programId);
    if (response?.data) {
      setProgramLevels(response.data.reverse());
    }
  }, []);

  const getPrograms = useCallback(async (programId: string) => {
    const response = await getProgramsById(programId);
    if (response?.data) {
      await getProgramsLevels(response.data.id);
      setSelectedProgram(response.data);
    }
  }, []);

  const getUserActivities = useCallback(async (userId: string) => {
    const response = await getUserActivitiesById(userId);
    if (response?.data) {
      setUserActivities(response.data);
    }
  }, []);

  const getUserProfile = useCallback(async (userId: string) => {
    const response = await getUserProfileById(userId);
    if (response?.data) {
      setSelectedUser(response.data);
      await getPrograms(response.data.programId);
      await getUserActivities(userId);
    }
  }, []);

  return (
    <MainView style={{ backgroundImage: `url(${backgroundImage})` }}>
      <LogoView>
        <LogoImage src={logo} />
      </LogoView>
      <ContentView>
        <UserRanking userList={usersState} onSelectUser={selectUser} />
        {selectedUser && selectedProgram && programLevels && (
          <UserProfile
            selectedUser={selectedUser}
            userProgram={selectedProgram}
            programLevels={programLevels}
          />
        )}
        {userActivities && <UserActivities userActivities={userActivities} />}
      </ContentView>
    </MainView>
  );
}
