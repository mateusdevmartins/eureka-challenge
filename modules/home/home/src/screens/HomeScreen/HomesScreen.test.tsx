import renderer from 'react-test-renderer';
import { HomeScreen } from './HomeScreen';

test('Avatar should match snapshot', () => {
  const component = renderer.create(<HomeScreen />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
