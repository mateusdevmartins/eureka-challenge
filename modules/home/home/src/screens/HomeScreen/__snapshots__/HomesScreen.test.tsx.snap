// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should match snapshot 1`] = `
<div
  className="styles__MainView-sc-1mf3dib-1 kjwRaq"
  style={
    Object {
      "backgroundImage": "url(undefined)",
    }
  }
>
  <div
    className="styles__LogoView-sc-1mf3dib-3 eZYvwz"
  >
    <img
      className="styles__LogoImage-sc-1mf3dib-0 iUukRh"
      src=""
    />
  </div>
  <div
    className="styles__ContentView-sc-1mf3dib-2 bDvcKl"
  >
    <div
      className="styles__OutterView-m7ssp0-0 cFOpin"
    >
      <div
        className="styles__InnerView-m7ssp0-1 zdZa-d"
      >
        <div
          className="styles__MainView-sc-9awpl6-0 fbRiBg"
        >
          <div
            className="styles__ListItemContainer-sc-9awpl6-1 ftIQND"
          >
            <h1
              className="styles__ColumnTitle-sc-9awpl6-2 hfbWlz"
              style={
                Object {
                  "width": "24px",
                }
              }
            >
              Pos.
            </h1>
            <h1
              className="styles__ColumnTitle-sc-9awpl6-2 hfbWlz"
            >
              Member
            </h1>
          </div>
          <h1
            className="styles__ColumnTitle-sc-9awpl6-2 hfbWlz"
          >
            Points
          </h1>
        </div>
      </div>
    </div>
  </div>
</div>
`;
